File: backend/app/schemas/superadmin.py
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.models import (
    ClubInvitationStatus,
    ClubMembershipRole,
    ClubMembershipStatus,
    ClubOnboardingState,
    ClubOnboardingStep,
)
from app.schemas.finance import AccountingExportProfileMappingConfig


class SuperadminClubCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    location: str = Field(min_length=1, max_length=255)
    timezone: str = Field(min_length=1, max_length=64)


class SuperadminClubSummary(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    location: str
    timezone: str
    active: bool
    onboarding_state: ClubOnboardingState
    onboarding_current_step: ClubOnboardingStep
    registry_status: Literal["active", "onboarding", "paused"]
    finance_ready: bool
    finance_profile_count: int
    active_assignment_count: int
    created_at: datetime
    updated_at: datetime


class SuperadminClubListResponse(BaseModel):
    items: list[SuperadminClubSummary]
    total_count: int


class OnboardingStepStatus(BaseModel):
    key: ClubOnboardingStep
    label: str
    status: Literal["complete", "current", "upcoming"]
    ready: bool


class SuperadminFinanceProfileSummary(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    target_system: str
    is_active: bool


class SuperadminFinanceSetupSummary(BaseModel):
    selected_accounting_profile_id: uuid.UUID | None
    selected_accounting_profile_name: str | None
    profile_count: int
    active_profile_count: int
    setup_complete: bool
    mapping_ready: bool
    profiles: list[SuperadminFinanceProfileSummary]


class SuperadminRulesSetupSummary(BaseModel):
    rule_set_count: int
    active_rule_set_count: int
    pricing_matrix_count: int
    active_pricing_matrix_count: int
    setup_complete: bool
    rule_sets: list["SuperadminRuleSetSummary"]
    pricing_matrices: list["SuperadminPricingMatrixSummary"]


class SuperadminRuleSetSummary(BaseModel):
    id: uuid.UUID
    name: str
    applies_to: str
    priority: int
    active: bool
    rule_count: int


class SuperadminPricingMatrixSummary(BaseModel):
    id: uuid.UUID
    name: str
    active: bool
    rule_count: int


class SuperadminModuleCatalogItem(BaseModel):
    key: str
    label: str
    description: str


class SuperadminModuleSetupSummary(BaseModel):
    enabled_module_keys: list[str]
    enabled_module_count: int
    setup_complete: bool
    available_modules: list[SuperadminModuleCatalogItem]


class SuperadminAssignedUserSummary(BaseModel):
    membership_id: uuid.UUID
    user_id: uuid.UUID
    person_id: uuid.UUID
    display_name: str
    email: str
    role: ClubMembershipRole
    status: ClubMembershipStatus
    is_primary: bool


class SuperadminAssignmentCandidate(BaseModel):
    user_id: uuid.UUID
    person_id: uuid.UUID
    display_name: str
    email: str


class SuperadminAssignmentCandidateListResponse(BaseModel):
    items: list[SuperadminAssignmentCandidate]
    total_count: int


class SuperadminClubOnboardingDetailResponse(BaseModel):
    club: SuperadminClubSummary
    progress_percent: int
    steps: list[OnboardingStepStatus]
    finance: SuperadminFinanceSetupSummary
    rules: SuperadminRulesSetupSummary
    modules: SuperadminModuleSetupSummary
    assignments: list[SuperadminAssignedUserSummary]


class SuperadminClubOnboardingUpdateRequest(BaseModel):
    action: Literal["save_draft", "complete_step", "return_to_previous_step"]
    acted_step: ClubOnboardingStep
    name: str | None = Field(default=None, min_length=1, max_length=255)
    location: str | None = Field(default=None, min_length=1, max_length=255)
    timezone: str | None = Field(default=None, min_length=1, max_length=64)
    preferred_accounting_profile_id: uuid.UUID | None = None
    enabled_module_keys: list[str] | None = None


class SuperadminClubStatusUpdateRequest(BaseModel):
    active: bool


class SuperadminClubAssignmentUpsertRequest(BaseModel):
    person_id: uuid.UUID
    role: ClubMembershipRole


class SuperadminClubAssignmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    membership_id: uuid.UUID
    club_id: uuid.UUID
    person_id: uuid.UUID
    role: ClubMembershipRole
    status: ClubMembershipStatus
    is_primary: bool


class SuperadminClubInvitationCreateRequest(BaseModel):
    email: str = Field(min_length=3, max_length=255)
    role: ClubMembershipRole


class SuperadminClubInvitationResponse(BaseModel):
    invitation_id: uuid.UUID
    club_id: uuid.UUID
    person_id: uuid.UUID
    membership_id: uuid.UUID
    linked_user_id: uuid.UUID | None
    email: str
    role: ClubMembershipRole
    status: ClubInvitationStatus
    membership_status: ClubMembershipStatus
    expires_at: datetime
    created_at: datetime
    accept_token: str | None = None


class SuperadminClubInvitationListResponse(BaseModel):
    items: list[SuperadminClubInvitationResponse]
    total_count: int


class SuperadminAccountingProfileSummary(BaseModel):
    id: uuid.UUID
    club_id: uuid.UUID
    club_name: str
    club_slug: str
    code: str
    name: str
    target_system: str
    is_active: bool
    mapping_config: AccountingExportProfileMappingConfig
    created_by_person_id: uuid.UUID
    created_at: datetime
    updated_at: datetime


class SuperadminAccountingProfileListResponse(BaseModel):
    profiles: list[SuperadminAccountingProfileSummary]
    total_count: int


class SuperadminAccountingProfileCreateRequest(BaseModel):
    club_id: uuid.UUID
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    target_system: str = Field(min_length=1, max_length=64)
    is_active: bool = True
    mapping_config: AccountingExportProfileMappingConfig


class SuperadminAccountingProfileActivationRequest(BaseModel):
    is_active: bool


class SuperadminAccountingProfileBindRequest(BaseModel):
    profile_id: uuid.UUID


class SuperadminAccountingTemplateColumnSample(BaseModel):
    values: list[str]


class SuperadminAccountingTemplateParseResponse(BaseModel):
    file_name: str
    headers_detected: list[str]
    headerless: bool
    suggested_target_system: str
    suggested_mapping: dict[str, str]
    sample_rows: list[SuperadminAccountingTemplateColumnSample]
    warnings: list[str]


class SuperadminAccountingSampleLayoutResponse(BaseModel):
    target_system: str
    file_name: str
    headerless: bool
    delimiter: str
    headers: list[str]
    sample_csv: str
    notes: list[str]

File: frontend/src/types/superadmin.ts
import type { MembershipRole, MembershipStatus } from "./session";

export type ClubOnboardingState =
  | "onboarding_started"
  | "data_pending"
  | "setup_in_progress"
  | "ready_for_go_live"
  | "live";

export type ClubOnboardingStep = "basic_info" | "finance" | "rules" | "modules";
export type ClubRegistryStatus = "active" | "onboarding" | "paused";
export type OnboardingStepStatus = "complete" | "current" | "upcoming";

export interface SuperadminClubSummary {
  id: string;
  name: string;
  slug: string;
  location: string;
  timezone: string;
  active: boolean;
  onboarding_state: ClubOnboardingState;
  onboarding_current_step: ClubOnboardingStep;
  registry_status: ClubRegistryStatus;
  finance_ready: boolean;
  finance_profile_count: number;
  active_assignment_count: number;
  created_at: string;
  updated_at: string;
}

export interface SuperadminClubListResponse {
  items: SuperadminClubSummary[];
  total_count: number;
}

export interface SuperadminFinanceProfileSummary {
  id: string;
  code: string;
  name: string;
  target_system: string;
  is_active: boolean;
}

export interface SuperadminFinanceSetupSummary {
  selected_accounting_profile_id: string | null;
  selected_accounting_profile_name: string | null;
  profile_count: number;
  active_profile_count: number;
  setup_complete: boolean;
  mapping_ready: boolean;
  profiles: SuperadminFinanceProfileSummary[];
}

export interface SuperadminRulesSetupSummary {
  rule_set_count: number;
  pricing_matrix_count: number;
  setup_complete: boolean;
}

export interface SuperadminModuleSetupSummary {
  enabled_module_keys: string[];
  setup_complete: boolean;
}

export interface SuperadminAssignedUserSummary {
  membership_id: string;
  user_id: string;
  person_id: string;
  display_name: string;
  email: string;
  role: MembershipRole;
  status: MembershipStatus;
  is_primary: boolean;
}

export interface SuperadminAssignmentCandidate {
  user_id: string;
  person_id: string;
  display_name: string;
  email: string;
}

export interface SuperadminAssignmentCandidateListResponse {
  items: SuperadminAssignmentCandidate[];
  total_count: number;
}

export interface SuperadminOnboardingStep {
  key: ClubOnboardingStep;
  label: string;
  status: OnboardingStepStatus;
  ready: boolean;
}

export interface SuperadminClubOnboardingDetail {
  club: SuperadminClubSummary;
  progress_percent: number;
  steps: SuperadminOnboardingStep[];
  finance: SuperadminFinanceSetupSummary;
  rules: SuperadminRulesSetupSummary;
  modules: SuperadminModuleSetupSummary;
  assignments: SuperadminAssignedUserSummary[];
}

export interface SuperadminClubCreateInput {
  name: string;
  location: string;
  timezone: string;
}

export interface SuperadminClubOnboardingUpdateInput {
  name?: string;
  location?: string;
  timezone?: string;
  onboarding_state?: ClubOnboardingState;
  onboarding_current_step?: ClubOnboardingStep;
  preferred_accounting_profile_id?: string | null;
}

export interface SuperadminClubAssignmentInput {
  person_id: string;
  role: Extract<MembershipRole, "club_admin" | "club_staff">;
}

export interface SuperadminClubAssignmentResponse {
  membership_id: string;
  club_id: string;
  person_id: string;
  role: Extract<MembershipRole, "club_admin" | "club_staff">;
  status: MembershipStatus;
  is_primary: boolean;
}
